### ISSUE TITLE
[Feature] Incapsulate exceptions

### ISSUE THREAD

**Issue by scaytrase** [lezhnev74/openapi-psr7-validator#14]
Currently not only library exceptions are being thrown while validation, but also `respect/validation` library exception comes to the end user like

```
All of the required rules must pass for 18497
```
(expecting `int[]` here, not `int`)

It's very hard to manually process these exceptions and enrich their messages with things like validation data path where the error occured and also make end user create complex try..catch blocks.

I suggest to rethrow these exceptions and thus totally isolate end user from the used validation library knownledge. This will make library usage easier and allows further refactoring without being bond to the validation library because of exception contracts.

Also, at the moment, `respect/validation` exceptions should be part of `validate` method signature, since they are thrown and should be catched

**Comment by lezhnev74** [comment]
Good point. This is indeed a deep topic.
- there are a lot of places where validation is used under the hood. Sometimes it is delegated to the PHP itself (hence `TypeError`), sometimes validation library is used (`Respect/Validation`)
- the goal of those checks is **to protect the logic from unexpected input**. It does it pretty well already.
- what your concern is about is **clarity, what exactly went wrong**. Here could be a couple of ways to bring more clarity to the package validation issues.

Some random thoughts from the top of my head:
1. Add meaningful messages to respect/validation assertions 
2. Replace that validation library with something more explicit about errors
3. You've suggested catching errors and converting them to some other exception, I don't think I can create exceptions for any error that could pop up.

What do you think?

**Comment by scaytrase** [comment]
> You've suggested catching errors and converting them to some other exception, I don't think I can create exceptions for any error that could pop up.

I hope there is not so many exception type that should be converted

> the goal of those checks is to protect the logic from unexpected input. It does it pretty well already.

As for validation library any input is expected ) but, well... invalid?

Coverting exceptions for me is all about wrapping Respect exception into library exception, not really doing 1-to-1 converstion, and adding some stuff like breadcrumbs

**Comment by lezhnev74** [comment]
Ok, let me meditate for a while on this. I am not really happy with this `All of the required rules must pass` as well.

> As for validation library any input is expected ) but, well... invalid?

Yeah, I am a bit too defensive in this regard. There are some assumptions in the package, like **openapi schemas should be valid** (otherwise the underlying `cebe/openapi-reader` should have thrown an exception. User's input in unpredictable and can be anything. That's why I add an extra layer of types and validation just in case, to protect the logic.

**Comment by scaytrase** [comment]
I've created a PR #24 which brings some exception consistency to `Schema` namespace (at the moment)
